src/main/scala/edu/ie3/osmogrid/io/input/pbf/PbfGuardian.scala

```diff
@@ -0,0 +1,200 @@
+  ): Behavior[PbfReaderMsg] = Behaviors.setup[PbfReaderMsg] { ctx =>
+    ctx.log.info(s"Start reading pbf file using $noOfActors actors ...")
+    val pool = Routers.pool(poolSize = noOfActors) {
+      Behaviors
+        .supervise(PbfWorker(highwayTags, buildingTags, landuseTags))
+        .onFailure[Exception](SupervisorStrategy.restart)
+    }
+    val router = ctx.spawn(pool, s"pbf-worker-pool-${UUID.randomUUID()}")
+    val pbfIS = new FileInputStream(pbfFile)
+
+    val workerResponseMapper: ActorRef[PbfWorker.Response] =
+      ctx.messageAdapter(rsp => WrappedPbfWorkerResponse(rsp))
+
+    idle(
+      StateData(
+        pbfIS,
+        BlobTupleIterator.fromPbf(pbfIS),
+        router,
+        workerResponseMapper
+      )
+    )
+  }
+
+  private def idle(stateData: StateData): Behavior[PbfReaderMsg] = Behaviors
+    .receive[PbfReaderMsg] { case (ctx, msg) =>
+      msg match {
+        case Run(filter, sender) =>
+          // start the reading process
+          val start = System.currentTimeMillis()
+          val noOfBlobs = readPbf(
+            stateData.blobIterator,
+            stateData.workerPool,
+            stateData.workerResponseMapper,
+            filter
+          )
+          ctx.log.debug(s"Reading $noOfBlobs blobs ...")
+          idle(
+            stateData.copy(
+              sender = Some(sender),
+              noOfBlobs = noOfBlobs,
+              startTime = start
+            )
+          )
+        case WrappedPbfWorkerResponse(response) =>
+          response match {
+            case PbfWorker.ReadSuccessful(additionalOsmoGridModel) =>
+              addOsmoGridModel(additionalOsmoGridModel, stateData, ctx)
+            case PbfWorker.ReadFailed(blobHeader, blob, filter, exception) =>
+              ctx.log.error("Error reading blob data.", exception)
+              stateData.sender.foreach(
+                _ !
+                  PbfReadFailed(
+                    PbfReadFailedException(
+                      "Error while reading pbf file from blob!"
+                    )
+                      .initCause(exception)
+                  )
+              )
+              stopAndCleanup(stateData, ctx)
+          }
+      }
+    }
+
+  private def stopAndCleanup(
+      stateData: StateData,
+      ctx: ActorContext[PbfReaderMsg]
+  ): Behavior[PbfReaderMsg] = {
+    ctx.log.info("Stopping .pbf file reading!")
+    stateData.pbfIS.close()
+    Behaviors.stopped // stops this actor and all its children
+  }
+
+  private def addOsmoGridModel(
+      additionalOsmoGridModel: OsmoGridModel,
+      stateData: StateData,
+      ctx: ActorContext[PbfReaderMsg]
+  ) = {
+
+    def status(stateData: StateData): Unit = {
+      val currentProgress =
+        (stateData.noOfResponses.toDouble / stateData.noOfBlobs) * 100d
+      val fivePercent = Math.round((stateData.noOfBlobs / 100d) * 5)
+      if (
+        currentProgress != 0d && stateData.noOfResponses.toDouble % fivePercent == 0
```

Comment:
Is it crucial that this works safely. Because I guess the second condition will only be hit "randomly".

Reply:
I think this should be improved, don't know a simple solution off the top of my head right now though.

Reply:
So, is it mission-critical or can we go with this solution first?